OpenRA.Mods.Common/Traits/BotModules/HarvesterBotModule.cs

```diff
@@ -13,58 +13,77 @@
+		public readonly WDist HarvesterEnemyAvoidanceRadius = WDist.FromCells(8);
+
+		public override object Create(ActorInitializer init) { return new HarvesterBotModule(init.Self, this); }
+	}
+
+	public class HarvesterBotModule : ConditionalTrait<HarvesterBotModuleInfo>, IBotModule
 	{
-		readonly HackyAI ai;
 		readonly World world;
-		readonly IPathFinder pathfinder;
-		readonly DomainIndex domainIndex;
-		readonly ResourceLayer resLayer;
-		readonly ResourceClaimLayer claimLayer;
+		readonly Predicate<Actor> unitCannotBeOrdered;
+		HackyAI ai;
+		IPathFinder pathfinder;
+		DomainIndex domainIndex;
+		ResourceLayer resLayer;
+		ResourceClaimLayer claimLayer;
+		List<Actor> harvesters = new List<Actor>();
+		int scanForIdleHarvestersTicks;
 
-		public AIHarvesterManager(HackyAI ai, Player p)
+		public HarvesterBotModule(Actor self, HarvesterBotModuleInfo info)
+			: base(info)
+		{
+			world = self.World;
+			unitCannotBeOrdered = a => a.Owner != self.Owner || a.IsDead || !a.IsInWorld;
+		}
+
+		string IBotModule.Name { get { return Info.Name; } }
+
+		void IBotModule.Activate(HackyAI ai)
 		{
 			this.ai = ai;
-			world = p.World;
 			pathfinder = world.WorldActor.Trait<IPathFinder>();
 			domainIndex = world.WorldActor.Trait<DomainIndex>();
 			resLayer = world.WorldActor.TraitOrDefault<ResourceLayer>();
 			claimLayer = world.WorldActor.TraitOrDefault<ResourceClaimLayer>();
+			scanForIdleHarvestersTicks = Info.ScanForIdleHarvestersInterval;
 		}
 
-		CPos FindNextResource(Actor actor, Harvester harv)
+		void IBotModule.Tick()
 		{
-			var locomotorInfo = actor.Info.TraitInfo<MobileInfo>().LocomotorInfo;
-
-			Func<CPos, bool> isValidResource = cell =>
-				domainIndex.IsPassable(actor.Location, cell, locomotorInfo) &&
-				harv.CanHarvestCell(actor, cell) &&
-				claimLayer.CanClaimCell(actor, cell);
+			if (IsTraitDisabled)
+				return;
 
-			var path = pathfinder.FindPath(
-				PathSearch.Search(world, locomotorInfo, actor, true, isValidResource)
-					.WithCustomCost(loc => world.FindActorsInCircle(world.Map.CenterOfCell(loc), ai.Info.HarvesterEnemyAvoidanceRadius)
-						.Where(u => !u.IsDead && actor.Owner.Stances[u.Owner] == Stance.Enemy)
-						.Sum(u => Math.Max(WDist.Zero.Length, ai.Info.HarvesterEnemyAvoidanceRadius.Length - (world.Map.CenterOfCell(loc) - u.CenterPosition).Length)))
-					.FromPoint(actor.Location));
+			if (resLayer == null || resLayer.IsResourceLayerEmpty)
+				return;
 
-			if (path.Count == 0)
-				return CPos.Zero;
+			if (--scanForIdleHarvestersTicks > 0)
+				return;
 
-			return path[0];
-		}
+			harvesters.RemoveAll(unitCannotBeOrdered);
+			scanForIdleHarvestersTicks = Info.ScanForIdleHarvestersInterval;
 
-		public void Tick(List<Actor> harvesters)
-		{
-			if (resLayer == null || resLayer.IsResourceLayerEmpty)
-				return;
+			// Find new harvesters
+			var newHarvesters = world.ActorsHavingTrait<Harvester>().Where(a => a.Owner == ai.Player && !harvesters.Contains(a));
```

Comment:
Possibly, but to be honest, I'd prefer to leave performance optimizations for later to avoid potentially introducing some regression that delays merging further.
I can add a TODO, though.

Reply:
> I can add a TODO, though.

Done.